DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorVertexSoA.cc

```diff
@@ -0,0 +1,122 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/Common/interface/Handle.h"
+// DQM Histograming
+#include "DQMServices/Core/interface/MonitorElement.h"
+#include "DQMServices/Core/interface/DQMEDAnalyzer.h"
+#include "DQMServices/Core/interface/DQMStore.h"
+#include "CUDADataFormats/Vertex/interface/ZVertexHeterogeneous.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+
+class SiPixelPhase1MonitorVertexSoA : public DQMEDAnalyzer {
+public:
+  explicit SiPixelPhase1MonitorVertexSoA(const edm::ParameterSet&);
+  ~SiPixelPhase1MonitorVertexSoA() override;
+  void bookHistograms(DQMStore::IBooker& ibooker, edm::Run const& iRun, edm::EventSetup const& iSetup) override;
+  void analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) override;
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  edm::EDGetTokenT<ZVertexHeterogeneous> tokenSoAVertex_;
+  edm::EDGetTokenT<reco::BeamSpot> tokenBeamSpot_;
+  std::string topFolderName_;
+  MonitorElement* hnVertex;
+  MonitorElement* hx;
+  MonitorElement* hy;
+  MonitorElement* hz;
+  MonitorElement* hchi2;
+  MonitorElement* hptv2;
+};
+
+//
+// constructors
+//
+
+SiPixelPhase1MonitorVertexSoA::SiPixelPhase1MonitorVertexSoA(const edm::ParameterSet& iConfig) {
+  tokenSoAVertex_ = consumes<ZVertexHeterogeneous>(iConfig.getParameter<edm::InputTag>("pixelVertexSrc"));
+  tokenBeamSpot_ = consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("beamSpotSrc"));
+  topFolderName_ = iConfig.getParameter<std::string>("TopFolderName");
+}
+
+SiPixelPhase1MonitorVertexSoA::~SiPixelPhase1MonitorVertexSoA() {
+  // do anything here that needs to be done at desctruction time
+  // (e.g. close files, deallocate resources etc.)
+  edm::LogInfo("SiPixelPhase1MonitorVertexSoA") << ">>> Destroy SiPixelPhase1MonitorVertexSoA ";
+}
+
+// -- Analyze
+//
+void SiPixelPhase1MonitorVertexSoA::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  auto const& vsoa = *(iEvent.get(tokenSoAVertex_).get());
+  int nVertices = vsoa.nvFinal;
+  auto bsHandle = iEvent.getHandle(tokenBeamSpot_);
+  float x0 = 0., y0 = 0., z0 = 0., dxdz = 0., dydz = 0.;
+  if (!bsHandle.isValid()) {
+    edm::LogWarning("PixelVertexProducer") << "No beamspot found. returning vertexes with (0,0,Z) ";
+  } else {
+    const reco::BeamSpot& bs = *bsHandle;
+    x0 = bs.x0();
+    y0 = bs.y0();
+    z0 = bs.z0();
+    dxdz = bs.dxdz();
+    dydz = bs.dydz();
+  }
+  for (int iv = 0; iv < nVertices; iv++) {
+    auto z = vsoa.zv[iv];
+    auto x = x0 + dxdz * z;
+    auto y = y0 + dydz * z;
+    z += z0;
+    hx->Fill(x);
+    hy->Fill(y);
+    hz->Fill(z);
+    if (vsoa.ndof[iv] != 0)
+      hchi2->Fill(vsoa.chi2[iv] / vsoa.ndof[iv]);
+    hptv2->Fill(vsoa.ptv2[iv]);
```

Comment:
what about track multiplicity?

Reply:
I am trying to understand if `vsoa.ndof[iv]` does the right thing.
Why 

https://github.com/cms-sw/cmssw/blob/c44be9acb12672c530d1ed454901e1435f60a2cc/CUDADataFormats/Vertex/interface/ZVertexSoA.h#L17

but 

https://github.com/cms-sw/cmssw/blob/c44be9acb12672c530d1ed454901e1435f60a2cc/CUDADataFormats/Vertex/interface/ZVertexSoA.h#L19

isn't `ndof` a property of the vertex ?

Reply:
because of 
```reused as workspace for the number of nearest neighbours FIXME```
so ndof IS a property of the vertex and == to number-of-associated-tracks-1

Reply:
so it is safe to use `vsoa.ndof[iv]`. @sroychow can you put it back? 
Also we can just use `ndof+1` instead of counting the tracks:

https://github.com/cms-sw/cmssw/blob/15b20da0db8c06b7f38f2a91a3b7930d6dbebbb1/DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorVertexSoA.cc#L95-L99

right?